src/main/kotlin/com/terraformation/backend/db/IdentifierGenerator.kt

```diff
@@ -1,65 +1,86 @@
 ) {
-  private val log = perClassLogger()
-
   /**
-   * Returns the next unused user-facing identifier. This is used in places where we need a unique
-   * value to identify a resource but it's not acceptable to display the underlying integer ID from
-   * the database, e.g., accession numbers.
-   *
-   * Identifiers are of the form YYYYMMDDXXX where XXX is a numeric suffix of three or more digits
-   * that starts at 000 for the first identifier created on a particular date. The desired behavior
-   * is for the suffix to represent the order in which entries were added to the system, so ideally
-   * we want to avoid gaps or out-of-order values, though it's fine for that to be best-effort.
-   *
-   * The implementation uses a database sequence. The sequence's values follow the same pattern as
-   * the identifiers, but the suffix is always 10 digits; it is rendered as a 3-or-more-digit value
-   * by this method.
+   * Generates a new identifier for an organization.
    *
-   * If the date part of the sequence value doesn't match the current date, this method resets the
-   * sequence to the zero suffix for the current date.
-   *
-   * Note that there is a bit of a race condition if multiple terraware-server instances happen to
-   * allocate their first identifier of a given day at the same time; they might both reset the
-   * sequence. To guard against that, [AccessionStore.create] will retry a few times if it gets a
-   * unique constraint violation on the accession number.
+   * @param timeZone Time zone to use to determine the current date.
    */
-  fun generateIdentifier(): String {
+  fun generateIdentifier(
+      organizationId: OrganizationId,
+      timeZone: ZoneId = ZoneOffset.UTC,
+  ): String {
     val suffixMultiplier = 10000000000L
-    val todayAsLong = LocalDate.now(clock).format(DateTimeFormatter.BASIC_ISO_DATE).toLong()
+    val todayAsLong =
+        LocalDate.ofInstant(clock.instant(), timeZone)
+            .format(DateTimeFormatter.BASIC_ISO_DATE) // "20221031"
+            .toLong()
+    val firstValueForToday = todayAsLong * suffixMultiplier
 
     val sequenceValue =
-        dslContext.select(ACCESSION_NUMBER_SEQ.nextval()).fetchOne(ACCESSION_NUMBER_SEQ.nextval())!!
-    val datePart = sequenceValue / suffixMultiplier
-    val suffixPart = sequenceValue.rem(suffixMultiplier)
-
-    val suffix =
-        if (todayAsLong != datePart) {
-          val firstValueForToday = todayAsLong * suffixMultiplier
+        with(IDENTIFIER_SEQUENCES) {
           dslContext
-              .alterSequence(ACCESSION_NUMBER_SEQ)
-              .restartWith(firstValueForToday + 1)
-              .execute()
-          log.info("Resetting identifier sequence to $firstValueForToday")
-          0
-        } else {
-          suffixPart
+              .insertInto(IDENTIFIER_SEQUENCES)
+              .set(ORGANIZATION_ID, organizationId)
+              .set(NEXT_VALUE, firstValueForToday)
+              .onDuplicateKeyUpdate()
```

Comment:
Does this perform the retry behavior?

Reply:
This gets translated to PostgreSQL-specific syntax, `ON CONFLICT (organization_id) DO UPDATE`. The organization ID is this table's primary key. The end effect is to insert the starting value if there isn't already a row for an organization, but if there is, run the update part of the query.

The database ensures this is all done atomically, so if two server instances try this at the same time, one will block until the other is done and only then will it read the current value and add 1 to it.

Reply:
ooh, that makes sense. thanks for the detailed explanation!